internal/workflow/replicator.go

```diff
@@ -106,26 +121,102 @@ func (w *Replicator) execute(ctx workflow.Context, request *ReplicatorRequest) e
+				newRequest.StartHeight = startHeight
+				logger.Info(
+					"checkpoint reached",
+					zap.Reflect("newRequest", newRequest),
+				)
+				return workflow.NewContinueAsNewError(ctx, w.name, &newRequest)
+			}
+			endHeight := startHeight + batchSize
 			if endHeight > request.EndHeight {
 				endHeight = request.EndHeight
 			}
-			_, err := w.replicator.Execute(ctx, &activity.ReplicatorRequest{
-				StartBlockHeight: startHeight,
-				EndBlockHeight:   endHeight,
-				Parallelism:      parallelism,
-				Tag:              tag,
-				UpdateWatermark:  request.UpdateWatermark,
-				Compression:      dataCompression,
-			})
-			if err != nil {
-				return xerrors.Errorf("failed to replicate blocks from %v-%v: %w", startHeight, endHeight, err)
+
+			wg := workflow.NewWaitGroup(ctx)
+			wg.Add(parallelism)
+			miniBatchCount := int((endHeight-startHeight-1)/miniBatchSize + 1)
+			inputChannel := workflow.NewNamedBufferedChannel(ctx, "replicator.input", miniBatchCount)
+			for batchStart := startHeight; batchStart < endHeight; batchStart = batchStart + miniBatchSize {
+				inputChannel.Send(ctx, batchStart)
+			}
+			inputChannel.Close()
+
+			reprocessChannel := workflow.NewNamedBufferedChannel(ctx, "replicator.reprocess", miniBatchCount)
+			defer reprocessChannel.Close()
+
+			// Phase 1: running mini batches in parallel.
+			for i := 0; i < parallelism; i++ {
+				workflow.Go(ctx, func(ctx workflow.Context) {
+					defer wg.Done()
+					for {
+						var batchStart uint64
+						if ok := inputChannel.Receive(ctx, &batchStart); !ok {
+							break
+						}
+						batchEnd := batchStart + miniBatchSize
+						if batchEnd > endHeight {
+							batchEnd = endHeight
+						}
+						_, err := w.replicator.Execute(ctx, &activity.ReplicatorRequest{
+							Tag:         tag,
+							StartHeight: batchStart,
+							EndHeight:   batchEnd,
+							Parallelism: parallelism,
+							Compression: dataCompression,
+						})
+						if err != nil {
+							reprocessChannel.Send(ctx, batchStart)
+							logger.Warn(
+								"queued for reprocessing",
+								zap.Uint64("batchStart", batchStart),
+								zap.Error(err),
+							)
+						}
+					}
+				})
+			}
+			wg.Wait(ctx)
+
+			// Phase 2: reprocess any failed batches sequentially.
+			// This should happen rarely (only if we over stress the cluster or the cluster itself was crashing).
+			for {
+				var batchStart uint64
+				if ok := reprocessChannel.ReceiveAsync(&batchStart); !ok {
+					break
+				}
+				batchEnd := batchStart + miniBatchSize
+				if batchEnd > endHeight {
+					batchEnd = endHeight
+				}
+				_, err := w.replicator.Execute(ctx, &activity.ReplicatorRequest{
+					Tag:         tag,
+					StartHeight: batchStart,
+					EndHeight:   batchEnd,
+					Parallelism: parallelism,
+					Compression: dataCompression,
+				})
+				if err != nil {
+					return xerrors.Errorf("failed to replicate block from %d to %d: %w", batchStart, batchEnd, err)
+				}
+			}
+
+			// Phase 3: update watermark
```

Comment:
I think we need to validate the output to ensure the blocks are contiguous before we update the watermark.  If we use the metadata from the downloaded blocks, then we don't need to call getBlocksByHeight during the update watermark, which would further reduce the chance of error in the corner case of a reorg.

Reply:
Added validation before updating watermark.